modules/network_synchronizer/networked_controller.cpp

```diff
@@ -0,0 +1,1112 @@
+		// Skip this until the first input arrive.
+		return;
+	}
+
+	node->get_inputs_buffer_mut().begin_read();
+	node->call("controller_process", p_delta);
+}
+
+bool is_remote_frame_A_older(const FrameSnapshotSkinny &p_snap_a, const FrameSnapshotSkinny &p_snap_b) {
+	return p_snap_a.id < p_snap_b.id;
+}
+
+void ServerController::receive_inputs(Vector<uint8_t> p_data) {
+
+	// The packet is composed as follow:
+	// - The following four bytes for the first input ID.
+	// - Array of inputs:
+	// |-- First byte the amount of times this input is duplicated in the packet.
+	// |-- inputs buffer.
+	//
+	// Let's decode it!
+
+	const int data_len = p_data.size();
+
+	int ofs = 0;
+
+	ERR_FAIL_COND(data_len < 4);
+	const uint32_t first_input_id = decode_uint32(p_data.ptr() + ofs);
+	ofs += 4;
+
+	uint64_t inserted_input_count = 0;
+
+	BitArray bit_array;
+	bit_array.resize_in_bytes(data_len);
+	copymem(bit_array.get_bytes_mut().ptrw(), p_data.ptr(), data_len);
+
+	// Contains the entire packet and in turn it will be seek to specific location
+	// so I will not need to copy chunk of the packet data.
+	PlayerInputsReference pir;
+	pir.set_inputs_buffer(bit_array);
+
+	while (ofs < data_len) {
+
+		ERR_FAIL_COND_MSG(ofs + 1 > data_len, "The arrived packet size doesn't meet the expected size.");
+		// First byte is used for the duplication count.
+		const uint8_t duplication = p_data[ofs];
+		ofs += 1;
+
+		// Validate input
+		pir.inputs_buffer.seek(ofs * 8);
+		const int input_size_in_bits = node->call("count_input_size", &pir);
+		// Pad to 8 bits.
+		const int input_size =
+				Math::ceil((static_cast<float>(input_size_in_bits)) / 8.0);
+		ERR_FAIL_COND_MSG(ofs + input_size > data_len, "The arrived packet size doesn't meet the expected size.");
+
+		// The input is valid, populate the buffer.
+		for (int sub = 0; sub <= duplication; sub += 1) {
+
+			const uint64_t input_id = first_input_id + inserted_input_count;
+			inserted_input_count += 1;
+
+			if (current_input_buffer_id != UINT64_MAX && current_input_buffer_id >= input_id)
+				continue;
+
+			FrameSnapshotSkinny rfs;
+			rfs.id = input_id;
+
+			const bool found = std::binary_search(
+					snapshots.begin(),
+					snapshots.end(),
+					rfs,
+					is_remote_frame_A_older);
+
+			if (!found) {
+				rfs.inputs_buffer.get_bytes_mut().resize(input_size);
+				copymem(
+						rfs.inputs_buffer.get_bytes_mut().ptrw(),
+						p_data.ptr() + ofs,
+						input_size);
+
+				snapshots.push_back(rfs);
+
+				// Sort the new inserted snapshot.
+				std::sort(snapshots.begin(), snapshots.end(), is_remote_frame_A_older);
+			}
+		}
+
+		// We can now advance the offset.
+		ofs += input_size;
+	}
+
+	ERR_FAIL_COND_MSG(ofs != data_len, "At the end was detected that the arrived packet has an unexpected size.");
+}
+
+int ServerController::notify_input_checked(uint64_t p_input_id) {
+	ERR_PRINT("The method `notify_input_checked` must not be called on server. Be sure why it happened.");
```

Comment:
```suggestion
	ERR_PRINT("The method `notify_input_checked` must not be called on the server.");
```

I don't think we should tell user's to investigate errors. Or otherwise, we'd need to add this to all errors :)

Reply:
Make, sense :) removed.